libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -323,23 +544,246 @@ static MQTTStatus_t addRecord( MQTTPubAckInfo_t * records,
 static void updateRecord( MQTTPubAckInfo_t * records,
                           size_t recordIndex,
                           MQTTPublishState_t newState,
                           bool shouldDelete )
 {
+    uint16_t packetId;
+
+    assert( records != NULL );
+
     if( shouldDelete == true )
     {
+        /* Mark the record as invalid. */
         records[ recordIndex ].packetId = MQTT_PACKET_ID_INVALID;
-        records[ recordIndex ].qos = MQTTQoS0;
-        records[ recordIndex ].publishState = MQTTStateNull;
     }
     else
     {
-        records[ recordIndex ].publishState = newState;
+        /* For QoS2 messages, in order to preserve the message ordering, when
+         * a PUBREC is received for an outgoing publish, the record should be
+         * moved to the last. This move will help preserve the order in which
+         * a PUBREL needs to be resent in case of a session reestablishment. */
+        if( newState == MQTTPubRelSend )
+        {
+            packetId = records[ recordIndex ].packetId;
+            /* Mark the record at current index as invalid. */
+            records[ recordIndex ].packetId = MQTT_PACKET_ID_INVALID;
+
+            /* Add it as a new record to the end. A deletion makes room for
+             * at least one more packet. So ignore return value here. */
+            ( void ) addRecord( records,
+                                MQTT_STATE_ARRAY_MAX_COUNT,
+                                packetId,
+                                records[ recordIndex ].qos,
+                                newState );
+        }
+        else
+        {
+            records[ recordIndex ].publishState = newState;
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static uint16_t stateSelect( const MQTTContext_t * pMqttContext,
+                             uint16_t searchStates,
+                             MQTTStateCursor_t * pCursor )
+{
+    uint16_t packetId = MQTT_PACKET_ID_INVALID;
+    uint16_t outgoingStates = 0U;
+    const MQTTPubAckInfo_t * records = NULL;
+    bool stateCheck = false;
+
+    assert( pMqttContext != NULL );
+    assert( searchStates != 0U );
+    assert( pCursor != NULL );
+
+    /* Create a bit map with all the outgoing publish states. */
+    UINT16_SET_BIT( outgoingStates, MQTTPublishSend );
+    UINT16_SET_BIT( outgoingStates, MQTTPubAckPending );
+    UINT16_SET_BIT( outgoingStates, MQTTPubRecPending );
+    UINT16_SET_BIT( outgoingStates, MQTTPubRelSend );
+    UINT16_SET_BIT( outgoingStates, MQTTPubCompPending );
+
+    /* Only outgoing publish records need to be searched. */
+    assert( ( outgoingStates & searchStates ) > 0U );
+    records = pMqttContext->outgoingPublishRecords;
+
+    while( *pCursor < MQTT_STATE_ARRAY_MAX_COUNT )
+    {
+        /* Check if any of the search states are present. */
+        stateCheck = UINT16_CHECK_BIT( searchStates, records[ *pCursor ].publishState ) ? true : false;
+
+        if( stateCheck == true )
```

Comment:
I think this could just be:
```suggestion
        if( UINT16_CHECK_BIT( searchStates, records[ *pCursor ].publishState ) )
```

Reply:
I would prefer it to be assigned to a variable for making the debugging easier.